### ISSUE TITLE
performance related to fps?

### ISSUE THREAD

**Issue by kilmarnock** [watts4speed/fast-dqn-caffe#7]
Hi!

You are saying:
... 1/2 million training steps in about 1 hour ...
As far as I can see, the gathering of the current screen (->PreprocessScreen()) is synchronized to the number of training steps. So to increase (my) environment speed, I have to 
a) raise the frames per second or
b) resend the current frame twice
Is this correct?   Which path did you choose, i. e. what does ale.getScreen do? How many fps does ale produce?
Is there a way to decouple fps and learning rate, f.e. by retraining the convolutional layers offline?

**Comment by watts4speed** [comment]
Hi Kilmarnock,

The basic process is to load the replay memory with new "experience" as it
occurs.  The replay memory is sampled randomly to feed the training of the
NN.  This will spread the training out over a large amount of experience
and not just the most recent experience and is one of the basic techniques
that make the algorithm work.

Depending on your H/W the NN training or the ALE simulation might be the
bottleneck.  From what I remember on my system with a Titan-X, the GPU was
only loaded to maybe 40%.  The bottleneck was on the CPU running ALE and
also preparing the data for each training step and then transfering to the
GPU for training.

For a:
You can raise the FPS, by say running multiple games at the same time which
might load the replay memory faster, but if the bottleneck becomes the CPU
time to transfer from the replay memory to the GPU memory and run a
training step then adding more games might not make sense.

For b:
Resending the same frame twice I wouldn't expect to help.  You could just
double the training rate to get the same result approximately.  The
fundamental process that's working is the accumulation of experience from
observing the state of a system, making a predicted best move, and the
providing feedback or reward for the predicted best move.  The system
gradually get better by trying stuff and seeing what happens, and as the
replay memory fills up it throws out older less accurate experience.  Any
effective acceleration would involve making that feedback process faster.

One speedup for example is to optimize the ProcessScreen() method so it
works faster.

the ALE.fps will run at I think a "virtual" 60FPS, but really depends on
the speed of the CPU and is not related to a real world clock.  Everything
has been tuned so as to handle some issues with how ALE works so I wouldn't
change much there, other than to run multiple games at once on separate
threads, but still that might not be the bottleneck on your system.

Ale.GetScreen() will from what I remember return an image of the current
game screen.

I think retraining the CONV layers offline might help, but that only makes
sense if you're maxing out your GPU back propagating the loss.  You still
have a batch size of 32 in the forward propagation step that has to run
through the entire NN in any case which might be the bottleneck.

Deepmind as work out a different architecture called Gorilla and has scaled
it over a large number of CPU.  You can find a paper on their site.

Hope this helps,
W4S

**Comment by kilmarnock** [comment]
Hi W4S,
thank you for your fast reply. It changes my view on the system which is / will be very helpful.

> The replay memory is sampled randomly to feed the training of the
> NN
> running multiple games at the same time which might load the replay memory faster

Does that imply that values in the replay memory are more-or-less tripples with action-reward-image content? Where is the episodic part in the learning process? The effect is "I did this -> that is the status".  Where is " I did this, that and that and it was better than that, this and that"?
Can I really stream multiple games in one replay memory, mixing the situations/states? Looks like.

> Depending on your H/W the NN training or the ALE simulation might be the
> bottleneck.

I try to write my own environment. I have an avatar which has a walking speed. I have to? synchronize the system to actions per meter so that 

a) number_of_frames / number_of_actions = constant
b) actions_per_meter = constant

With a) necessary to make it possible that each action gets assigned a correct status-after-action and b) necessary to have the same amount of movement per action. (If I do 5 actions per meter will have another learning outcome than 25 actions per meter)?

**Comment by watts4speed** [comment]
"Where is the episodic part in the learning process?"

The replay is loaded as a game plays which I think is where the episodic
information is accumulated.  The training of the NN randomly samples the
replay and from what I understand it learns how to make moves, based the
the experience of all the episodes that are saved in replay.  There's no
explicit episodic memory.  It's implicit in the experience accumulated in
the replay memory.  The NN learns for any given input state, what the best
move it.  I can't explain off the top of my head how the reward propagates
back to any specific state, it's more like a class in reinforcement
learning would explain that.

Yes its basically a triplet at the high level.
